File: android/app/src/main/java/app/notomorrow/feature/workout/WorkoutEditSupport.kt
package app.notomorrow.feature.workout

import app.notomorrow.data.relation.WorkoutWithExercises
import app.notomorrow.model.SetKind
import app.notomorrow.service.localizedName
import app.notomorrow.util.Fmt
import java.time.Instant
import java.time.LocalDate
import java.time.ZoneId
import java.time.temporal.ChronoUnit
import java.util.Locale
import kotlin.math.ceil
import kotlin.math.floor
import kotlin.math.max

/**
 * The pure half of the workout editor — `WorkoutEditSupport.swift`: the draft a finished workout is
 * edited in, every edit on it, and the set timeline a save writes back. None of it touches Room.
 */

// MARK: - Draft

/**
 * Unsaved copy of a finished workout for the editor — `WorkoutDraft`. Plain values only: nothing
 * reaches Room before Save, so Cancel is free.
 *
 * Ids of rows added in the editor are negative (handed out by the caller), so they never collide
 * with a saved row's id and `isDirty = draft != original` stays a plain comparison.
 */
data class WorkoutDraft(
    val name: String,
    val notes: String,
    /** Epoch millis. */
    val startedAt: Long,
    /** Milliseconds. Kept exact until the user steps it, so an untouched duration never moves a set's time. */
    val durationMs: Long,
    val exercises: List<ExerciseDraft>,
) {
    val endedAt: Long get() = startedAt + durationMs

    /** Exercise ids in the draft (the picker shows them as "In"). */
    val exerciseIds: Set<String> get() = exercises.map { it.exerciseId }.toSet()

    /** Save needs a start in the past and an end no later than a minute from now. */
    fun isTimeValid(now: Long): Boolean = startedAt <= now && endedAt <= now + FUTURE_GRACE_MS

    val canShorten: Boolean get() = durationMs > DURATION_MIN_MINUTES * MINUTE_MS
    val canLengthen: Boolean get() = durationMs < DURATION_MAX_MINUTES * MINUTE_MS

    // MARK: Editing

    /** Moves the start to [day], keeping the time of day (wall clock). */
    fun withDay(day: LocalDate, zone: ZoneId): WorkoutDraft {
        val start = Instant.ofEpochMilli(startedAt).atZone(zone)
        val delta = ChronoUnit.DAYS.between(start.toLocalDate(), day)
        if (delta == 0L) return this
        return copy(startedAt = start.plusDays(delta).toInstant().toEpochMilli())
    }

    /**
     * Moves the start to [hour]:[minute] on the same day (wall clock, so a DST night doesn't shift
     * it). Seconds stay, so picking the old time again gives back the exact start.
     */
    fun withTime(hour: Int, minute: Int, zone: ZoneId): WorkoutDraft {
        val start = Instant.ofEpochMilli(startedAt).atZone(zone)
        if (start.hour == hour && start.minute == minute) return this
        return copy(startedAt = start.withHour(hour).withMinute(minute).toInstant().toEpochMilli())
    }

    /** − / + on the duration: snaps to the 5-minute grid (52 min → 50 or 55), within 5 min … 12 h. */
    fun steppingDuration(direction: Int): WorkoutDraft {
        val step = DURATION_STEP_MINUTES.toDouble()
        val minutes = Fmt.roundHalfAwayFromZero(durationMs / MINUTE_MS.toDouble() * 100) / 100
        val snapped = if (direction > 0) (floor(minutes / step) + 1) * step else (ceil(minutes / step) - 1) * step
        val clamped = snapped.coerceIn(DURATION_MIN_MINUTES.toDouble(), DURATION_MAX_MINUTES.toDouble())
        return copy(durationMs = clamped.toLong() * MINUTE_MS)
    }

    /** Runs [change] on one set. */
    fun updatingSet(setId: Long, exerciseId: Long, change: (SetDraft) -> SetDraft): WorkoutDraft =
        copy(
            exercises = exercises.map { exercise ->
                if (exercise.id != exerciseId) {
                    exercise
                } else {
                    exercise.copy(sets = exercise.sets.map { if (it.id == setId) change(it) else it })
                }
            },
        )

    /**
     * Ticks or unticks a row. A tick that would not count can't be made (no new "0 × 0" sets):
     * `null` means it was refused. An untouched legacy 0-rep row unticked by mistake can be ticked
     * back ([SetDraft.isUntouchedCompleted]).
     */
    fun togglingDone(setId: Long, exerciseId: Long): WorkoutDraft? {
        val set = exercises.firstOrNull { it.id == exerciseId }?.sets?.firstOrNull { it.id == setId } ?: return this
        if (set.isLogged) return updatingSet(setId, exerciseId) { it.copy(isDone = false) }
        val ticked = set.copy(isDone = true)
        return if (ticked.isLogged) updatingSet(setId, exerciseId) { ticked } else null
    }

    /**
     * "+ Add set": copies the last row (a warm-up becomes a normal set). Added rows start ticked:
     * editing a past workout is logging after the fact. An empty one counts once it has reps.
     */
    fun addingSet(exerciseId: Long, newId: Long): WorkoutDraft = copy(
        exercises = exercises.map { exercise ->
            if (exercise.id != exerciseId) return@map exercise
            val last = exercise.sets.lastOrNull()
            val kind = last?.kind?.let { if (it == SetKind.Warmup) SetKind.Normal else it } ?: SetKind.Normal
            exercise.copy(
                sets = exercise.sets + SetDraft(
                    id = newId,
                    sourceId = null,
                    kind = kind,
                    weightKg = last?.weightKg ?: 0.0,
                    reps = last?.reps ?: 0,
                    isDone = true,
                    originalCompletedAt = null,
                ),
            )
        },
    )

    fun deletingSet(setId: Long, exerciseId: Long): WorkoutDraft = copy(
        exercises = exercises.map { exercise ->
            if (exercise.id != exerciseId) exercise else exercise.copy(sets = exercise.sets.filterNot { it.id == setId })
        },
    )

    fun removingExercise(exerciseId: Long): WorkoutDraft = copy(exercises = exercises.filterNot { it.id == exerciseId })

    /** Move up (-1) / down (+1) among the exercises. */
    fun movingExercise(exerciseId: Long, offset: Int): WorkoutDraft {
        val from = exercises.indexOfFirst { it.id == exerciseId }
        val to = from + offset
        if (from < 0 || to !in exercises.indices) return this
        val list = exercises.toMutableList()
        list[from] = exercises[to]
        list[to] = exercises[from]
        return copy(exercises = list)
    }

    /** An exercise added from the picker: one ticked row, prefilled from the last time it was done. */
    fun appendingExercise(
        id: Long,
        setId: Long,
        exerciseId: String,
        name: String,
        primaryMuscle: String?,
        restSeconds: Int,
        template: SetValue?,
    ): WorkoutDraft {
        if (exerciseId in exerciseIds) return this
        val set = SetDraft(
            id = setId,
            sourceId = null,
            kind = SetKind.Normal,
            weightKg = template?.weightKg ?: 0.0,
            reps = template?.reps ?: 0,
            isDone = true,
            originalCompletedAt = null,
        )
        return copy(
            exercises = exercises + ExerciseDraft(
                id = id,
                sourceId = null,
                exerciseId = exerciseId,
                name = name,
                primaryMuscle = primaryMuscle,
                restSeconds = restSeconds,
                sets = listOf(set),
            ),
        )
    }

    /** The lowest id in the draft, so the caller can hand out new (negative) ones below it. */
    val lowestId: Long
        get() = (exercises.map { it.id } + exercises.flatMap { ex -> ex.sets.map { it.id } }).minOrNull() ?: 0L

    companion object {
        /** Duration steps: 5 minutes, from 5 minutes to 12 hours. */
        const val DURATION_STEP_MINUTES = 5
        const val DURATION_MIN_MINUTES = 5
        const val DURATION_MAX_MINUTES = 720

        const val MINUTE_MS = 60_000L

        /** An end up to a minute ahead of now still saves (the clock moved while editing). */
        const val FUTURE_GRACE_MS = 60_000L
    }
}

/** `ExerciseDraft`. [id] is the `workout_exercise` row id, or negative for one added in the editor. */
data class ExerciseDraft(
    val id: Long,
    /** The `workout_exercise` row this came from; `null` when it was added in the editor. */
    val sourceId: Long?,
    val exerciseId: String,
    val name: String,
    val primaryMuscle: String?,
    val restSeconds: Int,
    val sets: List<SetDraft>,
)

/** `SetDraft`. [id] is the `set_entry` row id, or negative for one added in the editor. */
data class SetDraft(
    val id: Long,
    /** The `set_entry` row this came from; `null` when it was added in the editor. */
    val sourceId: Long?,
    val kind: SetKind,
    val weightKg: Double,
    val reps: Int,
    /**
     * Ticked. It only counts (✓, saved as completed) while it has reps: clearing the reps to retype
     * them does not lose the tick, and nothing is ever logged as "0 × 0".
     */
    val isDone: Boolean,
    /** `completedAt` as saved. Save remaps it when the start or the duration changes. */
    val originalCompletedAt: Long?,
    /** The row as saved, for a row that came from the workout; `null` for one added in the editor. */
    val saved: Saved? = null,
) {
    /** The values a row had when the editor opened — `SetDraft.Saved`. */
    data class Saved(val kind: SetKind, val weightKg: Double, val reps: Int, val isDone: Boolean)

    /**
     * Counts as completed: ticked with reps, or a row completed before the no-"0 × 0" rule (reps 0)
     * that the user has left exactly as it was. Saving any other edit (a rename) must not un-log it,
     * which could also take the day's attendance with it.
     */
    val isLogged: Boolean get() = isDone && (reps > 0 || isUntouchedCompleted)

    /** Completed when the editor opened and not changed since. */
    val isUntouchedCompleted: Boolean
        get() {
            val saved = saved ?: return false
            if (!saved.isDone || originalCompletedAt == null) return false
            return saved == Saved(kind, weightKg, reps, isDone)
        }
}

/** `WorkoutDraft(workout:)` — built from the sorted exercises and sets. */
fun WorkoutWithExercises.toDraft(locale: Locale): WorkoutDraft = WorkoutDraft(
    name = workout.name,
    notes = workout.notes,
    startedAt = workout.startedAt,
    durationMs = max(0L, (workout.endedAt ?: workout.startedAt) - workout.startedAt),
    exercises = sortedExercises.map { entry ->
        ExerciseDraft(
            id = entry.workoutExercise.id,
            sourceId = entry.workoutExercise.id,
            exerciseId = entry.workoutExercise.exerciseId.orEmpty(),
            name = entry.exercise?.localizedName(locale).orEmpty(),
            primaryMuscle = entry.exercise?.primaryMuscles?.firstOrNull(),
            restSeconds = entry.workoutExercise.restSeconds,
            sets = entry.sortedSets.map { set ->
                SetDraft(
                    id = set.id,
                    sourceId = set.id,
                    kind = set.kind,
                    weightKg = set.weightKg,
                    reps = set.reps,
                    isDone = set.isCompleted,
                    originalCompletedAt = set.completedAt,
                    saved = SetDraft.Saved(set.kind, set.weightKg, set.reps, set.isCompleted),
                )
            },
        )
    },
)

/**
 * One draft exercise as the set table renders it (`WorkoutEditExerciseSection`): numbered like the
 * active table (warm-ups don't count), no Previous, no current row, ✓ = [SetDraft.isLogged].
 */
fun ExerciseDraft.toSectionUi(): WorkoutExerciseUi {
    var number = 0
    val rows = sets.mapIndexed { index, set ->
        if (set.kind != SetKind.Warmup) number += 1
        SetRowUi(
            id = set.id,
            order = index,
            kind = set.kind,
            weightKg = set.weightKg,
            reps = set.reps,
            isCompleted = set.isLogged,
            number = number,
            previous = null,
            isCurrent = false,
        )
    }
    return WorkoutExerciseUi(
        id = id,
        exerciseId = exerciseId,
        name = name,
        primaryMuscle = primaryMuscle,
        restSeconds = restSeconds,
        setCount = sets.size,
        isDone = false,
        last = null,
        sets = rows,
    )
}

// MARK: - Timeline

/**
 * Where each done set lands in time after an edit — `WorkoutTimeline`. The records rule orders sets
 * by `completedAt`, so the save keeps the timeline consistent: existing times follow the workout's
 * new start and duration, new ones borrow a neighbour's.
 */
object WorkoutTimeline {

    /**
     * Order-preserving map of [t] from `[oldStart, oldEnd]` onto `[newStart, newEnd]`, clamped into
     * the new range. Untouched times come back exactly and a pure shift moves every set by the same
     * amount, so ties stay ties. Integer millis throughout; [t] is clamped into the old range before
     * scaling (the map is monotonic, so the result is the same) and the product never overflows.
     */
    fun remap(t: Long, oldStart: Long, oldEnd: Long, newStart: Long, newEnd: Long): Long {
        if (newStart == oldStart && newEnd == oldEnd) return t
        val oldSpan = oldEnd - oldStart
        val newSpan = newEnd - newStart
        val mapped = if (newSpan == oldSpan || oldSpan <= 0) {
            t + (newStart - oldStart)
        } else {
            newStart + (t.coerceIn(oldStart, oldEnd) - oldStart) * newSpan / oldSpan
        }
        return mapped.coerceIn(newStart, max(newStart, newEnd))
    }

    /**
     * `completedAt` for every logged row of the draft, by row id. A row that had a time gets it
     * remapped. A row without one (added, or ticked in the editor) takes the nearest earlier timed
     * row of its exercise, else the nearest later one (ties then resolve by row order, as the
     * records rule does); an exercise with no timed row takes the latest time of the exercises
     * above it, else the start.
     */
    fun completedTimes(draft: WorkoutDraft, oldStart: Long, oldEnd: Long): Map<Long, Long> {
        val newStart = draft.startedAt
        val newEnd = draft.endedAt
        val times = LinkedHashMap<Long, Long>()
        var latestAbove: Long? = null
        for (exercise in draft.exercises) {
            val anchors: List<Long?> = exercise.sets.map { set ->
                val original = set.originalCompletedAt
                if (!set.isLogged || original == null) null else remap(original, oldStart, oldEnd, newStart, newEnd)
            }
            exercise.sets.forEachIndexed { index, set ->
                if (!set.isLogged) return@forEachIndexed
                val time = anchors[index]
                    ?: anchors.subList(0, index).lastOrNull { it != null }
                    ?: anchors.subList(index + 1, anchors.size).firstOrNull { it != null }
                    ?: latestAbove
                    ?: newStart
                times[set.id] = time
                latestAbove = max(latestAbove ?: time, time)
            }
        }
        return times
    }
}

File: NoTomorrow/Features/Workout/WorkoutEditSupport.swift
import Foundation
import SwiftData

extension Notification.Name {
    /// A finished workout was edited or deleted. Screens that cache derived workout data reload.
    static let workoutHistoryDidChange = Notification.Name("nt.workoutHistoryDidChange")
}

// MARK: - Draft

/// Unsaved copy of a finished workout for the editor. Plain values only: nothing reaches SwiftData before Save,
/// so Cancel is free and the exercise picker saving its own context can't commit half an edit.
struct WorkoutDraft: Equatable {
    var name: String
    var notes: String
    var startedAt: Date
    /// Seconds. Kept exact until the user steps it, so an untouched duration never moves a set's time.
    var duration: TimeInterval
    var exercises: [ExerciseDraft]

    var endedAt: Date { startedAt.addingTimeInterval(duration) }

    /// Duration steps: 5 minutes, from 5 minutes to 12 hours.
    static let durationStep = 5
    static let durationRange = 5...720

    /// Exercise ids in the draft (the picker shows them as "In").
    var exerciseIDs: Set<String> { Set(exercises.map(\.exerciseID)) }

    /// Save needs a start in the past and an end no later than a minute from now.
    func isTimeValid(now: Date = .now) -> Bool {
        startedAt <= now && endedAt <= now.addingTimeInterval(60)
    }

    var canShorten: Bool { duration > Double(Self.durationRange.lowerBound * 60) }
    var canLengthen: Bool { duration < Double(Self.durationRange.upperBound * 60) }
}

struct ExerciseDraft: Identifiable, Equatable {
    let id: UUID
    /// The `WorkoutExercise` this came from; nil when it was added in the editor.
    var sourceID: PersistentIdentifier?
    var exerciseID: String
    var name: String
    var primaryMuscle: String?
    var restSeconds: Int
    var sets: [SetDraft]
}

struct SetDraft: Identifiable, Equatable {
    /// The values a row had when the editor opened.
    struct Saved: Equatable {
        var kind: SetKind
        var weightKg: Double
        var reps: Int
        var isDone: Bool
    }

    let id: UUID
    /// The `SetEntry` this came from; nil when it was added in the editor.
    var sourceID: PersistentIdentifier?
    var kind: SetKind
    var weightKg: Double
    var reps: Int
    /// Ticked. It only counts (✓, saved as completed) while it has reps: clearing the reps to retype them
    /// does not lose the tick, and nothing is ever logged as "0 × 0".
    var isDone: Bool
    /// `completedAt` as saved. Save remaps it when the start or the duration changes.
    var originalCompletedAt: Date?
    /// The row as saved, for a row that came from the workout; nil for one added in the editor.
    var saved: Saved? = nil

    /// Counts as completed: ticked with reps, or a row completed before the no-"0 × 0" rule (reps 0) that the user
    /// has left exactly as it was. Saving any other edit (a rename) must not un-log it, which could also take the
    /// day's attendance with it.
    var isLogged: Bool { isDone && (reps > 0 || isUntouchedCompleted) }

    /// Completed when the editor opened and not changed since.
    var isUntouchedCompleted: Bool {
        guard let saved, saved.isDone, originalCompletedAt != nil else { return false }
        return saved == Saved(kind: kind, weightKg: weightKg, reps: reps, isDone: isDone)
    }
}

extension WorkoutDraft {
    init(workout: Workout) {
        name = workout.name
        notes = workout.notes
        startedAt = workout.startedAt
        duration = max(0, (workout.endedAt ?? workout.startedAt).timeIntervalSince(workout.startedAt))
        exercises = workout.sortedExercises.map { entry in
            ExerciseDraft(
                id: UUID(), sourceID: entry.persistentModelID,
                exerciseID: entry.exercise?.id ?? "", name: entry.exercise?.localizedName ?? "",
                primaryMuscle: entry.exercise?.primaryMuscles.first, restSeconds: entry.restSeconds,
                sets: entry.sortedSets.map { set in
                    SetDraft(id: UUID(), sourceID: set.persistentModelID, kind: set.kind, weightKg: set.weightKg,
                             reps: set.reps, isDone: set.isCompleted, originalCompletedAt: set.completedAt,
                             saved: .init(kind: set.kind, weightKg: set.weightKg, reps: set.reps, isDone: set.isCompleted))
                })
        }
    }

    // MARK: Editing

    /// Moves the start to `day`, keeping the time of day.
    mutating func setDay(_ day: Date, calendar: Calendar = .current) {
        let delta = calendar.dateComponents([.day], from: calendar.startOfDay(for: startedAt),
                                            to: calendar.startOfDay(for: day)).day ?? 0
        guard delta != 0, let moved = calendar.date(byAdding: .day, value: delta, to: startedAt) else { return }
        startedAt = moved
    }

    /// Moves the start to `time`'s hour and minute on the same day (wall clock, so a DST night doesn't shift it).
    /// Seconds stay, so picking the old time again gives back the exact start.
    mutating func setTime(_ time: Date, calendar: Calendar = .current) {
        let new = calendar.dateComponents([.hour, .minute], from: time)
        let old = calendar.dateComponents([.hour, .minute, .second], from: startedAt)
        guard let hour = new.hour, let minute = new.minute, hour != old.hour || minute != old.minute,
              let from = calendar.date(bySettingHour: old.hour ?? 0, minute: old.minute ?? 0, second: old.second ?? 0,
                                       of: startedAt),
              let to = calendar.date(bySettingHour: hour, minute: minute, second: old.second ?? 0, of: startedAt)
        else { return }
        startedAt = startedAt.addingTimeInterval(to.timeIntervalSince(from))
    }

    /// − / + on the duration: snaps to the 5-minute grid (52 min → 50 or 55), within 5 min … 12 h.
    mutating func stepDuration(by direction: Int) {
        let step = Double(Self.durationStep)
        let minutes = (duration / 60 * 100).rounded() / 100
        let snapped = direction > 0 ? (floor(minutes / step) + 1) * step : (ceil(minutes / step) - 1) * step
        let clamped = min(max(snapped, Double(Self.durationRange.lowerBound)), Double(Self.durationRange.upperBound))
        duration = clamped * 60
    }

    /// Runs `change` on one set.
    mutating func updateSet(_ setID: UUID, in exerciseID: UUID, _ change: (inout SetDraft) -> Void) {
        guard let e = exercises.firstIndex(where: { $0.id == exerciseID }),
              let s = exercises[e].sets.firstIndex(where: { $0.id == setID }) else { return }
        change(&exercises[e].sets[s])
    }

    /// Ticks or unticks a row. A tick that would not count can't be made (no new "0 × 0" sets): returns false. An
    /// untouched legacy 0-rep row unticked by mistake can be ticked back.
    @discardableResult
    mutating func toggleDone(_ setID: UUID, in exerciseID: UUID) -> Bool {
        var accepted = true
        updateSet(setID, in: exerciseID) { set in
            if set.isLogged {
                set.isDone = false
            } else {
                var ticked = set
                ticked.isDone = true
                if ticked.isLogged { set.isDone = true } else { accepted = false }
            }
        }
        return accepted
    }

    /// "+ Add set": copies the last row (a warm-up becomes a normal set). Added rows start ticked: editing a past
    /// workout is logging after the fact. An empty one counts once it has reps.
    mutating func addSet(to exerciseID: UUID) {
        guard let e = exercises.firstIndex(where: { $0.id == exerciseID }) else { return }
        let last = exercises[e].sets.last
        let kind: SetKind = last.map { $0.kind == .warmup ? .normal : $0.kind } ?? .normal
        exercises[e].sets.append(SetDraft(id: UUID(), sourceID: nil, kind: kind, weightKg: last?.weightKg ?? 0,
                                          reps: last?.reps ?? 0, isDone: true, originalCompletedAt: nil))
    }

    mutating func deleteSet(_ setID: UUID, in exerciseID: UUID) {
        guard let e = exercises.firstIndex(where: { $0.id == exerciseID }) else { return }
        exercises[e].sets.removeAll { $0.id == setID }
    }

    mutating func removeExercise(_ exerciseID: UUID) {
        exercises.removeAll { $0.id == exerciseID }
    }

    /// Move up (-1) / down (+1) among the exercises.
    mutating func moveExercise(_ exerciseID: UUID, by offset: Int) {
        guard let from = exercises.firstIndex(where: { $0.id == exerciseID }) else { return }
        let to = from + offset
        guard exercises.indices.contains(to) else { return }
        exercises.swapAt(from, to)
    }

    /// An exercise added from the picker: one ticked row, prefilled from the last time it was done.
    mutating func appendExercise(id: String, name: String, primaryMuscle: String?, restSeconds: Int,
                                 template: (weightKg: Double, reps: Int)?) {
        guard !exerciseIDs.contains(id) else { return }
        let set = SetDraft(id: UUID(), sourceID: nil, kind: .normal, weightKg: template?.weightKg ?? 0,
                           reps: template?.reps ?? 0, isDone: true, originalCompletedAt: nil)
        exercises.append(ExerciseDraft(id: UUID(), sourceID: nil, exerciseID: id, name: name,
                                       primaryMuscle: primaryMuscle, restSeconds: restSeconds, sets: [set]))
    }

    /// Row number in the Set column: warm-ups don't count (same as the active table).
    func setNumber(of setID: UUID, in exerciseID: UUID) -> Int {
        guard let exercise = exercises.first(where: { $0.id == exerciseID }) else { return 0 }
        var n = 0
        for set in exercise.sets {
            if set.kind != .warmup { n += 1 }
            if set.id == setID { break }
        }
        return n
    }
}

// MARK: - Timeline

/// Where each done set lands in time after an edit. The records rule orders sets by `completedAt`, so the save keeps
/// the timeline consistent: existing times follow the workout's new start and duration, new ones borrow a neighbour's.
enum WorkoutTimeline {

    /// Order-preserving map of `t` from `[oldStart, oldEnd]` onto `[newStart, newEnd]`, clamped into the new range.
    /// Untouched times come back exactly and a pure shift moves every set by the same amount, so ties stay ties.
    static func remap(_ t: Date, oldStart: Date, oldEnd: Date, newStart: Date, newEnd: Date) -> Date {
        if newStart == oldStart && newEnd == oldEnd { return t }
        let oldSpan = oldEnd.timeIntervalSince(oldStart)
        let newSpan = newEnd.timeIntervalSince(newStart)
        let mapped: Date
        if newSpan == oldSpan || oldSpan <= 0 {
            mapped = t.addingTimeInterval(newStart.timeIntervalSince(oldStart))
        } else {
            mapped = newStart.addingTimeInterval(t.timeIntervalSince(oldStart) * newSpan / oldSpan)
        }
        return min(max(mapped, newStart), max(newStart, newEnd))
    }

    /// `completedAt` for every logged row of the draft. A row that had a time gets it remapped. A row without one
    /// (added, or ticked in the editor) takes the nearest earlier timed row of its exercise, else the nearest later
    /// one (ties then resolve by row order, as the records rule does); an exercise with no timed row takes the
    /// latest time of the exercises above it, else the start.
    static func completedTimes(for draft: WorkoutDraft, oldStart: Date, oldEnd: Date) -> [UUID: Date] {
        let newStart = draft.startedAt
        let newEnd = draft.endedAt
        var times: [UUID: Date] = [:]
        var latestAbove: Date?
        for exercise in draft.exercises {
            let anchors: [Date?] = exercise.sets.map { set in
                guard set.isLogged, let original = set.originalCompletedAt else { return nil }
                return remap(original, oldStart: oldStart, oldEnd: oldEnd, newStart: newStart, newEnd: newEnd)
            }
            for (index, set) in exercise.sets.enumerated() where set.isLogged {
                let time = anchors[index]
                    ?? anchors[..<index].compactMap { $0 }.last
                    ?? anchors[(index + 1)...].compactMap { $0 }.first
                    ?? latestAbove
                    ?? newStart
                times[set.id] = time
                latestAbove = max(latestAbove ?? time, time)
            }
        }
        return times
    }
}
